pkg/cloudprovider/instance.go

```diff
@@ -388,6 +433,30 @@ func filterInstanceTypes(instanceTypes []*cloudprovider.InstanceType) []*cloudpr
 	return instanceTypes
 }
 
+func instancesFromOutput(out *ec2.DescribeInstancesOutput) ([]*ec2.Instance, error) {
+	if len(out.Reservations) == 0 {
+		return nil, cloudprovider.NewMachineNotFoundError(fmt.Errorf("instance not found"))
+	}
+	instances := lo.Flatten(lo.Map(out.Reservations, func(r *ec2.Reservation, _ int) []*ec2.Instance {
+		if len(r.Instances) > 0 {
+			return r.Instances
+		}
+		return nil
+	}))
+	instances = lo.Reject(instances, func(i *ec2.Instance, _ int) bool {
+		return aws.StringValue(i.State.Name) == ec2.InstanceStateNameTerminated ||
+			aws.StringValue(i.State.Name) == ec2.InstanceStateNameShuttingDown
```

Comment:
is there a reason you're doing this here rather than filtering on the instance states you actually want back in the Describe call?